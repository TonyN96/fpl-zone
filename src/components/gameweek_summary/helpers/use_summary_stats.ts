import { useContext } from "react";
import { AppDataContext } from "app_content";
import { GetPlayerById, numberWithCommas } from "helpers";
import { AppData } from "types";

import { SummaryStatType } from "../gameweek_summary";

export const useSummaryStats = (): SummaryStatType[] => {
  const { gameweeks, players } = useContext(AppDataContext) as AppData;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const currentGameweek = gameweeks.find((gw) => gw.is_current)!;

  const starPlayerId = currentGameweek.top_element_info?.id;
  const starPlayer = starPlayerId ? GetPlayerById(starPlayerId, players) : null;

  const mostTransferredIn = currentGameweek.most_transferred_in
    ? GetPlayerById(currentGameweek.most_transferred_in, players)
    : null;

  const mostCaptained = currentGameweek.most_captained
    ? GetPlayerById(currentGameweek?.most_captained, players)
    : null;

  const mostViceCaptained = currentGameweek.most_vice_captained
    ? GetPlayerById(currentGameweek.most_vice_captained, players)
    : null;

  const mostTransferredInCount = mostTransferredIn
    ? numberWithCommas(mostTransferredIn?.transfers_in_event)
    : null;

  const starPlayerScore = currentGameweek.top_element_info
    ? `${currentGameweek.top_element_info?.points} pts`
    : "N/A";

  return ([
    {
      label: "highest score",
      value: `${currentGameweek.highest_score} pts`
    },
    {
      label: "average score",
      value: `${currentGameweek.average_entry_score} pts`
    },
    {
      label: "star player",
      teamCode: starPlayer?.team_code,
      playerName: starPlayer?.web_name || "",
      value: starPlayerScore
    },
    {
      label: "most captained",
      teamCode: mostCaptained?.team_code,
      playerName: mostCaptained ? mostCaptained.web_name : "N/A"
    },
    {
      label: "most vice-captained",
      teamCode: mostViceCaptained?.team_code,
      playerName: mostViceCaptained ? mostViceCaptained.web_name : "N/A"
    },
    {
      label: "most transferred in",
      teamCode: mostTransferredIn?.team_code,
      playerName: mostTransferredIn?.web_name || "",
      value: mostTransferredInCount || ""
    }
  ]);
};
