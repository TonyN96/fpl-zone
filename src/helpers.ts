import { Gameweek, Position, Team } from "types";
import { Player } from "types/player";

export const GAME_STATUS_VALUES = {
  GAME_UPDATING: "Game is updating..",
  SEASON_FINISHED: "The current FPL season has finished, check back next season!",
  GAME_OK: "GAME_OK"
};

// Getting a player by their id
export const GetPlayerById = (playerId: number, players: Player[]): Player => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return players.find((player) => player.id === playerId)!;
};

// Getting a team by their id
export const getTeamById = (teamId: number, teams: Team[]): Team => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return teams.find((team) => team.id === teamId)!;
};

// Getting a player by their id
export const getPositionById = (positionId: number, positions: Position[]): Position => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return positions.find((pos) => pos.id === positionId)!;
};

export const getPlayerImageUrl = (player?: Player): string => {
  if (!player) {
    return `${process.env.PUBLIC_URL}/assets/images/player-placeholder.png`;
  }

  const imgId = player.photo.replace(".jpg", "");

  return `https://resources.premierleague.com/premierleague/photos/players/110x140/p${imgId}.png`;
};

export const getNormalizedString = (input: string): string => {
  return input
    .normalize("NFD")
    .replace(/[\u0300-\u036F]/g, "")
    .toUpperCase();
};

// Adding commas to large numbers
// https://stackoverflow.com/questions/2901102/how-to-print-a-number-with-commas-as-thousands-separators-in-javascript
export const numberWithCommas = (x: number): string => {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
};

// Formatting date to desired format
export const formatDate = (kickOffDateTime: Date): string => {
  return (
    `${kickOffDateTime.toLocaleDateString(navigator.language, { day: "numeric", month: "short" })
    } ${
      kickOffDateTime.toLocaleTimeString(navigator.language, { hour: "2-digit", minute: "2-digit" })}`
  );
};

/*
  The following function is used to check two things:

  1. If the the most recent gameweek deadline was was within the last 1.5 hours
    (If so, FPL data is innacurate as the game is being updated for approx 1.5 hours after
    each gameweek deadline)

  2. If there is no current gameweek then the season is finished - this must be
    relayed back to the user so they understand why data is unavailable
*/
export const checkGameStatus = (gameweeks: Gameweek[]): string => {
  const currentDateTime = new Date();

  const currentGameweek = gameweeks.find(gameweek => gameweek.is_current);
  const nextGameweek = gameweeks.find(gameweek => gameweek.is_next);

  if (!currentGameweek) {
    return GAME_STATUS_VALUES.SEASON_FINISHED;
  }

  /*
    Manually checking if next gamewek deadline has passed as there
    can sometimes be a delay in update of gameweek 'is_next' status
  */
  const relevantGameweek = !nextGameweek || new Date(nextGameweek.deadline_time) > currentDateTime
    ? currentGameweek
    : nextGameweek;

  const gameweekDeadline = new Date(relevantGameweek.deadline_time);
  const timeDifference = currentDateTime.getTime() - gameweekDeadline.getTime();

  // If current time is 1.5 hours (5400000 ms) since deadline, game will be updating
  if (timeDifference < 5400000) {
    return GAME_STATUS_VALUES.GAME_UPDATING;
  }

  return GAME_STATUS_VALUES.GAME_OK;
};

// Adding delay before next function call
export const delay = async (ms: number | undefined): Promise<void> => {
  await new Promise((res) => setTimeout(res, ms));
};
