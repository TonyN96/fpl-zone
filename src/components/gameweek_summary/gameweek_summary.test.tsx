import React from "react";
import { render, screen } from "@testing-library/react";
import GameweekSummary from "./gameweek_summary";
import { mockGameweek, mockPlayers } from "../../test/fixture-data";
import { GetPlayerById, numberWithCommas } from "helpers";
import "@testing-library/jest-dom/extend-expect";

describe("Gameweek Summary Tests", () => {
  const createComponent = (): JSX.Element => {
    return <GameweekSummary gameweek={mockGameweek} players={mockPlayers} />;
  };

  it("Renders as expected", () => {
    const { asFragment } = render(createComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  describe("Data displays correctly", () => {
    it("Highest Score", () => {
      render(createComponent());
      const label = "highest score";

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByTestId(`${label}-stat-value`)).toHaveTextContent(
        mockGameweek.highest_score.toString()
      );
    });

    it("Average score", () => {
      render(createComponent());
      const label = "average score";

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByTestId(`${label}-stat-value`)).toHaveTextContent(
        mockGameweek.average_entry_score.toString()
      );
    });

    it("Star player", () => {
      render(createComponent());
      const label = "star player";
      const starPlayer = GetPlayerById(mockGameweek.top_element, mockPlayers);
      const url = `${process.env.PUBLIC_URL}/assets/images/crests/${starPlayer.team_code}.png`;

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByAltText(`${label}-crest-img`)).toHaveAttribute("src", url);
      expect(screen.getByTestId(`${label}-player-name`)).toHaveTextContent(
        `${starPlayer.first_name} ${starPlayer.second_name}`
      );
      expect(screen.getByTestId(`${label}-stat-value`)).toHaveTextContent(
        starPlayer.event_points.toString()
      );
    });

    it("Most Captained", () => {
      render(createComponent());
      const label = "most captained";
      const mostCaptainedPlayer = GetPlayerById(mockGameweek.most_captained, mockPlayers);
      const url = `${process.env.PUBLIC_URL}/assets/images/crests/${mostCaptainedPlayer.team_code}.png`;

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByAltText(`${label}-crest-img`)).toHaveAttribute("src", url);
      expect(screen.getByTestId(`${label}-player-name`)).toHaveTextContent(
        `${mostCaptainedPlayer.first_name} ${mostCaptainedPlayer.second_name}`
      );
    });

    it("Most Vice-captained", () => {
      render(createComponent());
      const label = "most vice-captained";
      const mostViceCaptainedPlayer = GetPlayerById(mockGameweek.most_vice_captained, mockPlayers);
      const url = `${process.env.PUBLIC_URL}/assets/images/crests/${mostViceCaptainedPlayer.team_code}.png`;

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByAltText(`${label}-crest-img`)).toHaveAttribute("src", url);
      expect(screen.getByTestId(`${label}-player-name`)).toHaveTextContent(
        `${mostViceCaptainedPlayer.first_name} ${mostViceCaptainedPlayer.second_name}`
      );
    });

    it("Most Transferred In", () => {
      render(createComponent());
      const label = "most transferred in";
      const mostTransferredInPlayer = GetPlayerById(mockGameweek.most_transferred_in, mockPlayers);
      const url = `${process.env.PUBLIC_URL}/assets/images/crests/${mostTransferredInPlayer.team_code}.png`;

      expect(screen.getByTestId(`${label}-label`)).toHaveTextContent(label.toUpperCase());
      expect(screen.getByAltText(`${label}-crest-img`)).toHaveAttribute("src", url);
      expect(screen.getByTestId(`${label}-player-name`)).toHaveTextContent(
        `${mostTransferredInPlayer.first_name} ${mostTransferredInPlayer.second_name}`
      );
      expect(screen.getByTestId(`${label}-stat-value`)).toHaveTextContent(
        numberWithCommas(mostTransferredInPlayer.transfers_in_event)
      );
    });
  });
});
