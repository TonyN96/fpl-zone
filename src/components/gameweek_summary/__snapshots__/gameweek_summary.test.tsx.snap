// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameweek Summary Tests Renders as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-jdlf87"
  >
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="highest score-label"
      >
        HIGHEST SCORE
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="highest score-player-name"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="highest score-stat-value"
      >
        104 pts
      </p>
    </div>
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="average score-label"
      >
        AVERAGE SCORE
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="average score-player-name"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="average score-stat-value"
      >
        18 pts
      </p>
    </div>
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="star player-label"
      >
        STAR PLAYER
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <img
          alt="star player-crest-img"
          height="80%"
          src="/assets/images/crests/3.png"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="star player-player-name"
        >
          Jankin
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="star player-stat-value"
      >
        17 pts
      </p>
    </div>
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="most captained-label"
      >
        MOST CAPTAINED
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <img
          alt="most captained-crest-img"
          height="80%"
          src="/assets/images/crests/6.png"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="most captained-player-name"
        >
          Summerill
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="most captained-stat-value"
      />
    </div>
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="most vice-captained-label"
      >
        MOST VICE-CAPTAINED
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <img
          alt="most vice-captained-crest-img"
          height="80%"
          src="/assets/images/crests/7.png"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="most vice-captained-player-name"
        >
          Bauldrey
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="most vice-captained-stat-value"
      />
    </div>
    <div
      class="MuiBox-root css-1g4yje1"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-1f5yx6w-MuiTypography-root"
        data-testid="most transferred in-label"
      >
        MOST TRANSFERRED IN
      </h3>
      <div
        class="MuiBox-root css-jkw4gz"
      >
        <img
          alt="most transferred in-crest-img"
          height="80%"
          src="/assets/images/crests/7.png"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1e4rbu-MuiTypography-root"
          data-testid="most transferred in-player-name"
        >
          Triner
        </p>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1 css-19gyy9g-MuiTypography-root"
        data-testid="most transferred in-stat-value"
      >
        1,669 transfers in
      </p>
    </div>
  </div>
</DocumentFragment>
`;
